go/internal/feast/onlinestore/cassandraonlinestore.go

```diff
@@ -0,0 +1,299 @@
+	//store.clusterConfigs.Consistency = gocql.Quorum
+	store.clusterConfigs.Keyspace = keyspaceStr
+	loadBalancingPolicy, ok := onlineStoreConfig["load_balancing"]
+	if !ok {
+		loadBalancingPolicy = gocql.RoundRobinHostPolicy()
+		log.Warn().Msg("No load balancing policy selected; setting Round Robin Host Policy")
+	}
+	loadBalancingPolicyStr, _ := loadBalancingPolicy.(string)
+	if loadBalancingPolicyStr == "DCAwareRoundRobinPolicy" {
+		store.clusterConfigs.PoolConfig.HostSelectionPolicy = gocql.RoundRobinHostPolicy()
+	} else if loadBalancingPolicyStr == "TokenAwarePolicy(DCAwareRoundRobinPolicy)" {
+		// Configure fallback policy if unable to reach the shard
+		fallback := gocql.RoundRobinHostPolicy()
+		// If using ScyllaDB and setting this policy, this makes the driver shard aware to improve performance
+		store.clusterConfigs.PoolConfig.HostSelectionPolicy = gocql.TokenAwareHostPolicy(fallback)
+		if config.OnlineStore["type"] == "scylladb" {
+			store.clusterConfigs.Port = 19042
+		} else {
+			store.clusterConfigs.Port = 9042
+		}
+	}
+
+	store.clusterConfigs.Authenticator = gocql.PasswordAuthenticator{Username: usernameStr, Password: passwordStr}
+	createdSession, err := store.clusterConfigs.CreateSession()
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect to the ScyllaDB database")
+	}
+	store.session = createdSession
+	return &store, nil
+}
+
+func (c *CassandraOnlineStore) getFqTableName(tableName string) string {
+	return fmt.Sprintf(`"%s"."%s_%s"`, c.keyspace, c.project, tableName)
+}
+
+func (c *CassandraOnlineStore) getCQLStatement(tableName string, featureNames []string, nKeys int) string {
+	// TODO: Compare with multiple single-key concurrent queries like in the Python feature server
+	keyPlaceholders := make([]string, nKeys)
+	for i := 0; i < nKeys; i++ {
+		keyPlaceholders[i] = "?"
+	}
+
+	// this prevents fetching unnecessary features
+	quotedFeatureNames := make([]string, len(featureNames))
+	for i, featureName := range featureNames {
+		quotedFeatureNames[i] = fmt.Sprintf(`'%s'`, featureName)
+	}
+
+	return fmt.Sprintf(
+		`SELECT "entity_key", "feature_name", "event_ts", "value" FROM %s WHERE "entity_key" IN (%s) AND "feature_name" IN (%s)`,
+		tableName,
+		strings.Join(keyPlaceholders, ","),
+		strings.Join(quotedFeatureNames, ","),
+	)
+}
+
+func (c *CassandraOnlineStore) buildCassandraEntityKeys(entityKeys []*types.EntityKey) ([]interface{}, map[string]int, error) {
+	cassandraKeys := make([]interface{}, len(entityKeys))
+	cassandraKeyToEntityIndex := make(map[string]int)
+	for i := 0; i < len(entityKeys); i++ {
+		var key, err = utils.SerializeEntityKey(entityKeys[i], c.config.EntityKeySerializationVersion)
+		if err != nil {
+			return nil, nil, err
+		}
+		// encoding to hex
+		encodedKey := hex.EncodeToString(*key)
+		cassandraKeys[i] = encodedKey
+		cassandraKeyToEntityIndex[encodedKey] = i
+	}
+	return cassandraKeys, cassandraKeyToEntityIndex, nil
```

Comment:
Is it same as buildRedisEntityKeys? can we move to key_utils if its the same?

Reply:
Almost, but I chose to use the same hex key in the array and map here for consistency.